### ISSUE TITLE
Add a plug-in, it can auto switch the multi connections, so your don't need create multi connection items

### ISSUE THREAD

**Issue by allanpk716** [VShawn/PRemoteM#103]
**Is your feature request related to a problem? Please describe.**
It's not a problem. it's a new feature.
In some cases, I have a server, it can be connected in 2 ways ( 1. LAN.  2. FRP local host ), so i need create 2 connection items in PRemote and I need remember where am I then use which connection item to connect the server. 

**Describe the solution you'd like**

So I add a feature, that you can only create one connection item , use anywhere. see [Premote-Plugin-SmartProtocol](https://github.com/allanpk716/Premote-Plugin-SmartProtocol)

**Describe alternatives you've considered**
see [Premote-Plugin-SmartProtocol](https://github.com/allanpk716/Premote-Plugin-SmartProtocol)

**Additional context**
no

**Comment by majkinetor** [comment]
Couldn't this be already done with pre-connect command ?

**Comment by allanpk716** [comment]
> Couldn't this be already done with pre-connect command ?

No, PRemote need communicate with it. When PRemote get the value back and need change the IP and port to the best way then to connect. Of course, this's change only IP/PORT config in memory , does not affect the actual storage.

这个插件需要与 PRemote 交互，然后获取到最佳的返回值后，需要动态修改连接的 IP 和 Port。当然这个是内存级的，不会影响实际的存储，当次有效。

**Comment by majkinetor** [comment]
You could dynamically add entry in hosts file in pre connect script, although you would need to use the same port. Port   could be fixed tho by option of pre connect script to return back some data, and then it could also return any parameter connection. For example if powershell script is called, it could return something like

`@{ Address = '1.2.3.4'; Port = '8000 ' }` 

or if cmd is called, it could return that in local env vars.

Now, not sure if that is easier or not, but is definitely more general solution as you could literary do anything you want, including setting up VPN/FRP or whatever.

So, your solution (since you didn't give screenshots) is to add multiple IP/PORT settings to connection ? How would PRM know which one to use ? Any one that works ?

**Comment by allanpk716** [comment]
> You could dynamically add entry in hosts file in pre connect script, although you would need to use the same port. Port could be fixed tho by option of pre connect script to return back some data, and then it could also return any parameter connection. For example if powershell script is called, it could return something like
> 
> `@{ Address = '1.2.3.4'; Port = '8000 ' }`
> 
> or if cmd is called, it could return that in local env vars.
> 
> Now, not sure if that is easier or not, but is definitely more general solution as you could literary do anything you want, including setting up VPN/FRP or whatever.
> 
> So, your solution (since you didn't give screenshots) is to add multiple IP/PORT settings to connection ? How would PRM know which one to use ? Any one that works ?

see this page, will tell you how it worked. [Premote-Plugin-SmartProtocol](https://github.com/allanpk716/Premote-Plugin-SmartProtocol) :smile:

**Comment by majkinetor** [comment]
OK, thx:

> SPConfig.yaml

So addresses are in another config file. Complicated IMO :) 

As this is basically pre-processing of existing connection, I think we should go with script returning this. For example in Powershell that would be something like:


```powershell
Smart-Connection ($Options) {
    foreach ($option in $Options) {
        $ip, $port = $option -split ':'
        if (Test-Connection -TargetName $ip -TcpPort $port -TimeoutSeconds 5) {
            return $ip, $port
        }
    }
    throw 'No connection available'
}
```

![image](https://user-images.githubusercontent.com/85767/112981769-eda50b00-915b-11eb-8198-29a1351e2bee.png)

PRM could even allow specifying entire script in configuration (so that no external files are needed).

This solution is way more general of course and solves need for infinity number of plugins.

**Comment by allanpk716** [comment]
1. External script is too slow to do this things.
2. IMO,  *.ps1 script  is more hard to read and write, and not suitable for deal with this job, it's need multi threading.
3. It doesn't solve the problem of communicate with the main program and still requires additional coding.

Now it's just an extra configurator or interaction to do the setup.

And this pull request, does not affect the use of ordinary people, can wait for the subsequent interaction can be improved.

**Comment by majkinetor** [comment]
> External script is too slow to do this things.

Facts please, not wishful thinking: any script - PoweShell (with adequate params) or cmd can execute within 1s especially cmd which is probably almost instant but has very bad programming capabilities. 

Also, do you really need ultimate speed here ? Its not that different If you connect in 100ms or 3 seconds in non regular use case weird cases, what is important is easiness of setup and options.

> IMO, *.ps1 script is more hard to read and write, 

Really ? I provided 4 lines instead of your 100+ lines or so ?  

> and not suitable for deal with this job, it's need multi threading.

The above command support multithreading, as posh in general, see [example 1](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/test-connection?view=powershell-7.1#example-2--send-echo-requests-to-several-computers) or [example 2](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/test-connection?view=powershell-7.1#example-4--run-a-test-as-a-background-job).

> It doesn't solve the problem of communicate with the main program and still requires additional coding.

It solves the problem if script communicates back. It also solves 1M other problems.

To me this kind of feature is what categorizes as scratching ones itch. It bloats the main tool, requires maintenance and extra documentation and so on. I am not against having it, just not within main codebase , perhaps there should be external plugins repository. Then maintenance of this plugin isn't on main dev but community. Scripting extensions could be good for easy to use no-development-tools-needed cases as those come out of the box with operating system and don't require anything but notepad.

Of course, that is my personal opinion, please don't take this as something against your work which is nice idea. @VShawn might have totally other one.

**Comment by allanpk716** [comment]
I want know 1M other problem is？

**Comment by allanpk716** [comment]
You talk a lot, but you don't solve the problem of external execution and then changing the local IP and port.Don't just give that solution when you don't know what the requirements are.

**Comment by VShawn** [comment]
This implementation seems very controversial, and the merger is put on hold.

For me, this feature is dispensable. I created two profiles for the local network and public network of a server, and I didn't feel it troublesome in switching them. 

I dont like this implement with external tool, it‘s not friendly to config, and the address in external is not been encrypted, and it can not publish to Microsoft Store, but I don’t have a better idea, I once designed to allow a profile to contain multiple IP addresses and allow people to choose from the launcher menu, but this design would break the basic data structure in PRM, so I have not made my mind to change PRM like this.

**Comment by allanpk716** [comment]
Why don't you first think about what the solution is supposed to solve and then come up with a workable solution instead of just saying something that doesn't really matter
